New Channel Lifecycle documentation

tomczoink opened the pull request:
This adds lifecycle documentation, along with references to said documentation in relevant locations.

mattheworiordan commented on content/general/channel_lifecycle.textile:
```diff
@@ -0,0 +1,124 @@
+
+h2(#reactor). Reactor rules
+
+You can associate "reactor rules":https://www.ably.io/reactor with metachannels in the same way as for regular channels; configure these using the "Message":/realtime/messages source and use a @channelFilter@ that matches the qualified metachannel name, or simply the name @[meta]channel.lifecycle@. The @Channel Lifecycle@ webhook source is no longer supported with the channel lifecycle metachanne ...
```
Why is `Channel Lifecycle` in code syntax?  It's not an object is it?  If it is, why not link to that object?

mattheworiordan replied:
Let's not say no longer for webhooks, state they are not supported. Perhaps: "Please note: Channel lifecycle events are supported with Reactor Webhooks. "

mattheworiordan commented on content/general/channel_lifecycle.textile:
```diff
@@ -0,0 +1,124 @@
+
+h3(#occupancy). Occupancy
+
+- metrics := an optional dictionary of membership categories for a realtime channel and their counts. Membership categories include:
```
This is more of a question than anything else, but is it necessary to have three different types when the customer is really only ever seeing one type i.e. a JSON object with ChannelDetails, with a status, and an occupancy hash.  Could we not just show this as a single `ChannelDetails` object so that conceptually it's easier for a developer to udnerstand?  I am not sure this is the case, but it's a question worth consdiering./

mattheworiordan commented on content/general/channel_lifecycle.textile:
```diff
@@ -0,0 +1,124 @@
+
+h2(#types). Types
+
+The primary payload of lifecycle events for channels is the "ChannelDetails":#channel-details type which contains the @channelId@ and other static information about the channel, plus a @status@ containing a "ChannelStatus":#channel-status instance which contains information about the current state of the channel.
```
Is there another payload?  If not, isn't this just "The payload of lifecycle events..."

mattheworiordan commented on content/rest-api/index.textile:
```diff
@@ -534,6 +534,14 @@ bc[json]. [{
+
+<%= partial 'shared/_channel_metadata' %>
+
+h3(#enumeration-rest). Retrieve enumeration of all active channels
```
This looks odd.  Why not just "Enumerate all active channels" instead of retrieve enumeration?

mattheworiordan commented on content/rest/channels.textile:
```diff
@@ -182,10 +184,14 @@ h3(#namespaces). Channel namespaces
 
 Channels expose a <span lang="default">@presence@</span><span lang="csharp">@Presence@</span> member which a client can use to obtain present members and presence event history for the channel itself. See the "REST presence":/rest/presence documentation for details.
 
+h2(#channel-metadata). Channel Metadata
```
As mentioned elsewhere, I think we need to better position this, explain it, and use consistent namgin.

mattheworiordan commented on content/shared/_channel_enumeration.textile:
```diff
@@ -0,0 +1,22 @@
+The REST API now includes the endpoint:
```
Not now, this is living documentation.  The REST API includes an endpoint to enumerate active channels within your app.

mattheworiordan commented on content/shared/_channel_enumeration.textile:
```diff
@@ -0,0 +1,22 @@
+- prefix := optionally limits the query to only those channels whose name starts with the given prefix<br>__Type: @string@__
+- by := _value_ optionally specifies whether to return just channel names (@by=id@) or "ChannelDetails":general/channel_lifecycle#channel-details (@by=value@)
+
+The credentials presented with the request must include the @list-channels@ permission for the wildcard resource ('@*@') or, if a @prefix@ is specified, for a resource that matches the @prefix@.
```
@Srushtika please advise @tomczoink, this permission is not valid I bleieve.

mattheworiordan replied:
To be consistent with other APIs, I would like us to define the return values very clearly.

mattheworiordan commented on content/shared/_channel_enumeration.textile:
```diff
@@ -0,0 +1,22 @@
+
+The credentials presented with the request must include the @list-channels@ permission for the wildcard resource ('@*@') or, if a @prefix@ is specified, for a resource that matches the @prefix@.
+
+Client libraries do not currently support this API, but it is usable via the generic "request API":/rest/usage#request.
```
Client libraries do not currently provide a dedicated API to enumerate channels, but make this available using the "request" method.

Here is a trivial example:
...

mattheworiordan commented on content/shared/_channel_enumeration.textile:
```diff
@@ -0,0 +1,22 @@
+
+Client libraries do not currently support this API, but it is usable via the generic "request API":/rest/usage#request.
+
+@Enumeration@ is possible of all channels in an app, by repeated calls to the API, following the @next@ relative link on each successive call, until there is no @next@ relative link. However, the state of the app and the cluster itself can change during that enumeration. This API therefore has the following limitations:
```
Please note that the `next` link only applies to users accessing the raw REST API.  This is in most situations, not how people will do it given we just told themt o use the request method above.  I believe you may need to document the raw REST API and this API as part of the client library SDKs slightly differently.

mattheworiordan commented on content/shared/_channel_metadata.textile:
```diff
@@ -0,0 +1,11 @@
+
+Returns a "ChannelDetails":/general/channel_lifecycle#channel-details for the given channel, indicating global occupancy. A side-effect of this request, in this pre-release of this API, is that it will cause the channel in question to become activated; therefore it is primarily intended to be used in conjunction with the "enumeration API":#enumeration-rest or in situations where the application h ...
+
+The credentials presented with the request must include the @list-channels@ permission for the channel in question.
```
As above, this permission is wrong.

mattheworiordan commented on content/general/channel_lifecycle.textile:
```diff
@@ -0,0 +1,124 @@
+
+<%= partial 'shared/_channel_enumeration' %>
+
+h2(#reactor). Reactor rules
```
This shoudl be grouped with Subscribing.  Logically, we have two sets of APIs.  Realtie APIs and REST APIs.  Let's keep the logic together. Then for Realtime APIs we have events generated when channe lifecycle vents are generated, or if occupancy events are activated, we generate events periodically for as long as channels remain active.

Srushtika commented:
@mattheworiordan, sure I'll help @tomczoink with usability information.

tomczoink commented:
@mattheworiordan @Srushtika I have updated per feedback, and also updated the article https://support.ably.io/solution/articles/3000053844-how-does-ably-count-messages to contain info on channel lifecycles. Anything else to do or good to go?

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+h2(#overview). Overview
+
+The realtime "channel lifecycle":#lifecycle-events and "occupancy":#occupancy events are published onto "metachannels":#metachannels that can be subscribed to using our existing "realtime libraries":/realtime. This is often used in conjunction with the "Ably Reactor":/reactor, which allows you to send this data to your "serverless endpoints":/general/functions, "queues":general/queues, and "strea ...
```
Perhaps:

Every channel in Ably's Data Stream Network that carries various messages has a "lifecycle":todo-link. The "channel lifecycle events":#lifecycle- are triggered whenever a channel changes its "lifecycle state":todo-link. A channel also consists of various occupants, meaning the participants that could either be publishers, subscribers, presence members or presence subscribers. The "occupancy":#occupancy events are triggered whenever there is a change in the number of any of these participants. Both the channel lifecycle and the occupancy events are published onto certain special channels called the "metachannels":#metachannels, that can be subscribed to, using our existing "realtime libraries":/realtime. This is often used in conjunction with the "Ably Reactor":/reactor, which allows you to send this data to your "serverless endpoints":/general/functions, "queues":general/queues, and "streams":general/firehose.

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+The realtime "channel lifecycle":#lifecycle-events and "occupancy":#occupancy events are published onto "metachannels":#metachannels that can be subscribed to using our existing "realtime libraries":/realtime. This is often used in conjunction with the "Ably Reactor":/reactor, which allows you to send this data to your "serverless endpoints":/general/functions, "queues":general/queues, and "strea ...
+
+The "REST API":/rest-api can also be used "enumerate all active channels":/rest/channel-lifecycle#enumeration-rest, or obtain the status of an "individual channel":#metadata-rest. At present, REST "requests":/rest/usage#request for lifecycle events will not count towards your message limit. If you are making use of "Reactor":/reactor however, then each message sent over Reactor *will* count towar ...
```
Perhaps:

You can also make use of the "REST API":/rest-api, to enumerate all active channels..

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+h2(#use-cases). Use cases
+
+Having access to channel metadata can provide numerous benefits. In a scenario where the number of subscribers to a channel goes well beyond a hundred, usage of other options such as "presence":/core-features/presence become less effective. You could instead make use of our "occupancy events":#occupancy and "channel metadata":#metachannels to check the number of active subscribers, getting accura ...
```
'subscribers of a channel'?

Srushtika replied:
become -> becomes

Srushtika replied:
FYI - occupancy events are part of channel metadata. Did you mean channel enumeration?

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+Having access to channel metadata can provide numerous benefits. In a scenario where the number of subscribers to a channel goes well beyond a hundred, usage of other options such as "presence":/core-features/presence become less effective. You could instead make use of our "occupancy events":#occupancy and "channel metadata":#metachannels to check the number of active subscribers, getting accura ...
+
+Equally, you may want to only publish data when you know someone is subscribed for your data. Therefore, "channel lifecycle events":#lifecycle-events can notify you when a channel becomes active, and can notify you when the last subscriber leaves the channel.
```
Perhaps:

'Equally, you may want to only publish data when you know someone is subscribed for your data.' -> ' Equally, you may want to publish your data only if there is a subscriber for that data" ?

Srushtika replied:
The channel occupancy/lifecycle events only notify about those numbers or events so perhaps saying that a user will be notified when the last subscriber leaves is a bit misleading.

Maybe just say, you can set up notifications for when the last subscriber leaves the channel, using the occupancy updates?

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+h2(#tutorials). Tutorials
+
+Check out some of our "tutorials":/tutorials to see examples of using metachannels and channel lifecycles.
```
Perhaps:

If you wish to see an example implementation of the channel metadata, you can check out our "Channel Lifecycle Events":todo-link tutorial, "Channel Occupancy Events" tutorial or the "Channel Enumeration" tutorial.

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+Check out some of our "tutorials":/tutorials to see examples of using metachannels and channel lifecycles.
+
+Through a realtime connection you can subscribe to both "channel lifecycle":#lifecycle-events and "occupancy":#occupancy events. This is done through a channel's "metachannel":#metachannels.
```
'Channel's metachannel' doesn't really make sense. 

-> This is done using the metachannel as explained next.

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+h2(#lifecycle-events). Lifecycle events
+
+The following events arise as messages on the channel lifecycle metachannel. In all cases, the @data@ member of the message is a "ChannelDetails":#channel-details instance.
```
Perhaps instead of saying 'on the channel lifecycle metachannel' - give the actual channel name embedded in @@?

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+- channel.region.inactive := indicates that the channel has been deactivated in a specific region. The included @ChannelDetails.status@ includes "occupancy":#occupancy in that region
+- channel.occupancy := indicates a change in global "occupancy":#occupancy. Not all "occupancy":#occupancy changes are sent; there is a minimum interval on a per-channel basis, so if the "occupancy":#occupancy is changing continuously, then only periodic snapshots of occupancy are sent. Further roll-up of events may occur depending on the capacity of the lifecycle channel. Any "occupancy":#occupa ...
+
+It is worth noting that "occupancy":#occupancy is presently only available to "enterprise":https://www.ably.io/pricing/enterprise accounts and for testing. "Contact us":https://www.ably.io/contact to discuss your requirements if this is relevant to you.
```
Not too important but perhaps 'presently' -> 'currently' so it's consistent with our language throughout?

Srushtika replied:
Maybe: 

"only available to "enterprise":https://www.ably.io/pricing/enterprise accounts and for testing." --> "only available to our "enterprise":https://www.ably.io/pricing/enterprise customers. However, you can still try and test out the occupancy events using your free account."

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+It is worth noting that "occupancy":#occupancy is presently only available to "enterprise":https://www.ably.io/pricing/enterprise accounts and for testing. "Contact us":https://www.ably.io/contact to discuss your requirements if this is relevant to you.
+
+h3(#occupancy-realtime). Occupancy with realtime
```
Maybe just call it 'Occupancy Events' like the 'Lifecycle Events' above

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+
+h3(#occupancy-realtime). Occupancy with realtime
+
+You can subscribe to channel "occupancy":#occupancy events by attaching to a meta channel using a realtime connection. However, realtime connections and channels have rate limits, typically "15 for free accounts":#https://support.ably.io/solution/articles/3000053845-do-you-have-any-connection-message-rate-or-other-limits-on-accounts. If the rate of channel lifecycle events exceeds this, a realtim ...
```
Maybe specify the name of the metachannel exactly instead of saying 'a metachannel'?

Srushtika replied:
Maybe specify 15 per what i.e. 15/sec 15/month etc

Srushtika commented on content/realtime/channel-lifecycle.textile:
````diff
@@ -0,0 +1,141 @@
+  {"[meta]*":["subscribe"]}
+  ```
+
+  The above will allow for the key to @subscribe@ to any @meta@ channel. The wildcard @*@ indicates anything can follow the @[meta]@ claim, so an example of a valid channel would be @[meta]log@. However, this capability will not allow for any other actions to be performed on the metachannels, nor will it allow the key to do anything with any non-metachannels.
````
Remove whitespace in the beginning

Srushtika commented on content/realtime/channel-lifecycle.textile:
````diff
@@ -0,0 +1,141 @@
+  }
+  ```
+
+  The above permission provides two capabilities; the ability to perform any action on any metachannel (such as @[meta]log@) with @"[meta]*":["*"]@, and the ability to perform any action on any channel (such as @another:channel@) with @"*":["*"]@. However, you are *never* able to publish or be present in a metachannel, thus this permission in effect would result in an actual permission excluding  ...
````
Remove whitespace in the beginning

Srushtika commented on content/realtime/channel-lifecycle.textile:
````diff
@@ -0,0 +1,141 @@
+  }
+  ```
+
+  Although the above provide all capabilities in all regular channel, without a @[meta]@ permission being explicitily specified, you will be unable to peform any actions on a @[meta]@ channel.
````
Remove whitespace in the beginning

tomczoink replied:
All the whitespaces are there so minimize works.

Srushtika commented on content/realtime/channel-lifecycle.textile:
```diff
@@ -0,0 +1,141 @@
+h1. Channel Lifecycle API Reference
+
+inline-toc.
+  Types:
```
perhaps show the hierarchy of these objects with indentation of representing them as JSON nested objects in curly braces

Srushtika commented on content/realtime/channels.textile:
```diff
@@ -205,6 +207,10 @@ bc[swift]. realtime.channels.get("chatroom").attach { error in
 
 Clients attach to a channel in order to participate on that channel in any way (either to publish, subscribe, or be present on the channel).
 
+h2(#channel-metadata). Channel metadata
```
Let's discuss with Matt if we should call the other file channel metadata and not channel lifecycle as the former is more relevant

tomczoink commented on content/general/channel-lifecycle.textile:
```diff
@@ -0,0 +1,154 @@
+
+  The above will allow for the key to @subscribe@ to any @meta@ channel. The wildcard @*@ indicates anything can follow the @[meta]@ claim, so an example of a valid channel would be @[meta]log@. However, this capability will not allow for any other actions to be performed on the metachannels, nor will it allow the key to do anything with any non-metachannels.
+
+minimize. See an example of a capability allowing all permissible actions on all metachannels and all normal channels
```
This is resolved as part of #544
